Parsing an empty file produces a fatal error

Sending an empty file to the Python driver produces a fatal error (with by definition stops the driver from processing more requests). This shouldn't be so since empty files are common in Python (__init__.py), it should just produce an error with an empty UAST returned.

Oh, so cool! 🎉 

Just spend few hours of life banding my head on

```
Read kalipo-ir/harvester/__init__.py, 0 bytes	Parsing file:'kalipo-ir/harvester/__init__.py', content:''
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1186147]

goroutine 1 [running]:
github.com/bblfsh/sdk/uast.(*Node).ProtoSize(0x0, 0xc4201e0f90)
```

--- a/native/python_package/python_driver/requestprocessor.py
+++ b/native/python_package/python_driver/requestprocessor.py
@@ -22,14 +22,6 @@ RawRequest = NewType('RawRequest', Dict[AnyStr, Any])
 Request    = NewType('Request',    Dict[str, Any])
 Response   = NewType('Response',   Dict[AnyStr, Any])
 
-class RequestCheckException(Exception):
-    """
-    Exception produced while there is an error during the processing
-    of the request. It will cause an error reply to be produced on the
-    output buffer.
-    """
-    pass
-
 class EmptyCodeException(Exception):
     """
     Exception produced when the input code is empty. This should
@@ -75,26 +67,6 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         """
         pass
 
-    def _parse_input_request(self, request: RawRequest) -> Tuple[str, str]:
-        """
-        Check the incoming request package and validate that the 'content' and
-        'language' keys are not missing and that 'language' and 'language_version'
-        are the right ones for this driver. It will also call _preprare_dict
-        to covnvert the request bytestrings to str.
-
-        :param request: The incoming request, already deserialized.
-
-        .. raises::
-            EmptyCodeException if the code was empty.
-        """
-        str_request = self._tostr_request(request)
-        code = asstr(str_request.get('content', ''))
-
-        if not code:
-            raise EmptyCodeException('Bad input message, missing content')
-
-        return code, asstr(str_request.get('filepath', ''))
-
     @abc.abstractmethod
     def _send_response(self, response: Response) -> None:
         """
@@ -118,18 +90,7 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         :param status: error type, 'error' or 'fatal'
         """
 
-        if status == 'fatal':
-            ret_ast = None
-        elif ast:
-            ret_ast = ast
-        else:
-            # Empty modules are still modules
-            ret_ast = {"PY3AST": {
-                    "ast_type"   : "Module",
-                    "lineno"     : 1,
-                    "col_offset" : 1,
-                  }}
-
+        ret_ast = None if status == 'fatal' else ast
         logging.error('Filepath: {}, Errors: {}'.format(filepath, self.errors))
         response = Response({
             'status': status,
@@ -156,24 +117,35 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         self.errors = []
 
         try:
-            code, filepath = self._parse_input_request(request)
+            str_request = self._tostr_request(request)
+            code = asstr(str_request.get('content', ''))
 
-            # We want the code detection to be fast and we prefer Python3 AST so using
-            # the stop_on_ok_ast will avoid running a Python2 subprocess to check the
-            # AST with Python2 if the Python3 version (which is a better AST anyway) worked
-            resdict  = detector.detect(codestr=code, stop_on_ok_ast=True)
-            codeinfo = resdict['<code_string>']
-            version  = codeinfo['version']
+            if code:
+                # We want the code detection to be fast and we prefer Python3 AST so using
+                # the stop_on_ok_ast will avoid running a Python2 subprocess to check the
+                # AST with Python2 if the Python3 version (which is a better AST anyway) worked
+                resdict  = detector.detect(codestr=code, stop_on_ok_ast=True)
+                codeinfo = resdict['<code_string>']
+                version  = codeinfo['version']
 
-            if version in (3, 6) and codeinfo['py3ast']:
-                ast = codeinfo['py3ast']
-            elif version in (1, 2) and codeinfo['py2ast']:
-                ast = codeinfo['py2ast']
+                if version in (3, 6) and codeinfo['py3ast']:
+                    ast = codeinfo['py3ast']
+                elif version in (1, 2) and codeinfo['py2ast']:
+                    ast = codeinfo['py2ast']
+                else:
+                    raise Exception('Could not determine Python version')
+
+                if not ast:
+                    raise Exception('Empty AST generated from non empty code')
             else:
-                raise Exception('Could not determine Python version')
-
-            if not ast:
-                raise Exception('Empty AST generated')
+                # Module with empty code (like __init__.py) return a module-only AST
+                # since this would still have semantic meaning for Python
+                ast = {"PY3AST": {
+                        "ast_type"   : "Module",
+                        "lineno"     : 1,
+                        "col_offset" : 1,
+                       }}
+                version = 3
 
             response = Response({
                 'status'           : 'ok',
@@ -185,6 +157,7 @@ class RequestProcessor(metaclass=abc.ABCMeta):
                     'driver'           : 'python23:%s' % __version__,
                 }
             })
+
             if filepath:
                 response['filepath'] = filepath
 

--- a/native/python_package/python_driver/requestprocessor.py
+++ b/native/python_package/python_driver/requestprocessor.py
@@ -4,7 +4,7 @@ import logging
 from pydetector import detector
 from traceback import format_exc
 from python_driver.version import __version__
-from typing import (Any, IO, NewType, Tuple, cast, List, Iterator, Dict)
+from typing import (Any, IO, NewType, Tuple, cast, List, Iterator, Dict, Optional)
 
 # typing.AnyStr is bugged on this version of MyPy, so...:
 AnyStr = Any
@@ -27,6 +27,14 @@ class RequestCheckException(Exception):
     Exception produced while there is an error during the processing
     of the request. It will cause an error reply to be produced on the
     output buffer.
+    """
+    pass
+
+class EmptyCodeException(Exception):
+    """
+    Exception produced when the input code is empty. This should
+    generate an error (because we can't parse anything) but not
+    a fatal one (e. g. empty __init__.py files)
     """
     pass
 
@@ -67,7 +75,7 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         """
         pass
 
-    def _check_input_request(self, request: RawRequest) -> Tuple[str, str]:
+    def _parse_input_request(self, request: RawRequest) -> Tuple[str, str]:
         """
         Check the incoming request package and validate that the 'content' and
         'language' keys are not missing and that 'language' and 'language_version'
@@ -77,13 +85,13 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         :param request: The incoming request, already deserialized.
 
         .. raises::
-            RequestCheckException if the request failed to validate.
+            EmptyCodeException if the code was empty.
         """
         str_request = self._tostr_request(request)
         code = asstr(str_request.get('content', ''))
 
         if not code:
-            raise RequestCheckException('Bad input message, missing content')
+            raise EmptyCodeException('Bad input message, missing content')
 
         return code, asstr(str_request.get('filepath', ''))
 
@@ -99,7 +107,8 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         """
         pass
 
-    def _return_error(self, filepath: AnyStr='', status: AnyStr='error') -> None:
+    def _return_error(self, filepath: AnyStr='', status: AnyStr='error',
+            ast: Optional[Dict[Any, Any]] = None) -> None:
         """
         Build and send to stdout and error response. Also log
         the errors to the python_driver.log.
@@ -109,11 +118,24 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         :param status: error type, 'error' or 'fatal'
         """
 
+        if status == 'fatal':
+            ret_ast = None
+        elif ast:
+            ret_ast = ast
+        else:
+            # Empty modules are still modules
+            ret_ast = {"PY3AST": {
+                    "ast_type"   : "Module",
+                    "lineno"     : 1,
+                    "col_offset" : 1,
+                  }}
+
         logging.error('Filepath: {}, Errors: {}'.format(filepath, self.errors))
         response = Response({
             'status': status,
             'errors': self.errors,
             'driver': 'python23:%s' % __version__,
+            'ast': ret_ast,
         })
         if filepath:
             response['filepath'] = filepath
@@ -134,7 +156,7 @@ class RequestProcessor(metaclass=abc.ABCMeta):
         self.errors = []
 
         try:
-            code, filepath = self._check_input_request(request)
+            code, filepath = self._parse_input_request(request)
 
             # We want the code detection to be fast and we prefer Python3 AST so using
             # the stop_on_ok_ast will avoid running a Python2 subprocess to check the
@@ -167,6 +189,10 @@ class RequestProcessor(metaclass=abc.ABCMeta):
                 response['filepath'] = filepath
 
             self._send_response(response)
+
+        except EmptyCodeException:
+            self.errors.append('Code field empty')
+            self._return_error(filepath, status='error', ast=ast)
 
         except:
             status = 'fatal' if ast is None else 'error'

--- a/native/python_package/test/test_python_driver.py
+++ b/native/python_package/test/test_python_driver.py
@@ -18,7 +18,7 @@ sys.path.append('..')
 from python_driver import __version__, get_processor_instance
 from python_driver.requestprocessor import (
     Request, Response, RequestProcessorJSON,
-    InBuffer, RequestCheckException
+    InBuffer, RequestCheckException, EmptyCodeException
 )
 
 if TEST_MSGPACK:
@@ -251,7 +251,7 @@ class Test20ReqProcMethods(TestPythonDriverBase):
             self._restart_data('json')
             brequest = convert_bytes(self.data, to_bytes=True)
             processor = RequestProcessorMSGPack(self.recvbuffer)
-            res = processor._check_input_request(brequest)
+            res = processor._parse_input_request(brequest)
             self.assertEqual(res[1], 'test.py')
 
         def test_20_check_input_bad(self) -> None:
@@ -259,8 +259,8 @@ class Test20ReqProcMethods(TestPythonDriverBase):
             del self.data['content']
             brequest = convert_bytes(self.data, to_bytes=True)
             processor = RequestProcessorMSGPack(self.recvbuffer)
-            with self.assertRaises(RequestCheckException) as _:  # noqa: F841
-                processor._check_input_request(brequest)
+            with self.assertRaises(EmptyCodeException) as _:  # noqa: F841
+                processor._parse_input_request(brequest)
 
         def test_30_send_response_msgpack(self) -> None:
             self._restart_data('msgpack')
@@ -290,6 +290,7 @@ class Test20ReqProcMethods(TestPythonDriverBase):
         self.assertDictEqual(res[0] , {'driver': 'python23:%s' % __version__,
                                        'errors': ['test error'],
                                        'filepath': 'test.py',
+                                       'ast': None,
                                        'status': 'fatal'})
 
 

--- a/native/python_package/test/test_python_driver.py
+++ b/native/python_package/test/test_python_driver.py
@@ -17,9 +17,7 @@ else:
 sys.path.append('..')
 from python_driver import __version__, get_processor_instance
 from python_driver.requestprocessor import (
-    Request, Response, RequestProcessorJSON,
-    InBuffer, RequestCheckException, EmptyCodeException
-)
+    Request, Response, RequestProcessorJSON, InBuffer, EmptyCodeException)
 
 if TEST_MSGPACK:
     from python_driver.requestprocessor import RequestProcessorMSGPack

